1433: changed look of password reset blade, added password validation hints

avas opened the pull request:
Changes made for issue #1433:
* Added `IPasswordCheckService` interface implemented by `PasswordCheckService` class that is tested by `PasswordCheckServiceTests`. It validates password strength and returns `PasswordValidationResult` that helps to decide:
    * Is the password valid? Can it be used as account password?
    * If not, what rules it violates?
* `SecurityController`: added `ValidatePasswordAsync` method that invokes password validation (using `IPasswordCheckService`). It always returns `Ok(passwordValidationResult)`, regardless of password validation status.
* Changed look of the password reset blade: 
    * Removed input for password re-typing;
    * Added switch for resetting user password when first logging in (it doesn't work yet, but required for future parts of this issue).
    * Added password strength validation. Now the blade calls `SecurityController.ValidatePasswordAsync()` on each password change and, if the password is not valid, displays error messages.

Note that requests from blade to `SecurityController.ValidatePasswordAsync()` are not throttled. This method does not do any heavy-processing like database queries or other service requests (at least for now), so it *probably* should be OK.

avas commented on VirtoCommerce.Platform.Data.Security/SecurityService.cs:
```diff
@@ -266,7 +266,7 @@ public virtual async Task<SecurityResult> ChangePasswordAsync(string name, strin
         }
 
-        public virtual async Task<SecurityResult> ResetPasswordAsync(string name, string newPassword)
+        public virtual async Task<SecurityResult> ResetPasswordAsync(string name, string newPassword, bool forcePasswordChange)
```
Well, I added it for future work on issue #1433. But I agree, it is not used yet and isn't really related to the purpose of this PR. Removed it.

tatarincev commented on VirtoCommerce.Platform.Web/Security/AuthenticationOptionsBasedPasswordCheckOptions.cs:
```diff
@@ -0,0 +1,39 @@
+    /// Implementation of <see cref="IPasswordCheckService"/> that gets password validation options from
+    /// <see cref="AuthenticationOptions"/>.
+    /// </summary>
+    public class AuthenticationOptionsBasedPasswordCheckOptions : IPasswordCheckOptions
```
I guess better will be using already exists `AuthenticationOptions` type instead of this decorator

tatarincev commented on VirtoCommerce.Platform.Web/Scripts/app/security/blades/account-resetPassword.js:
```diff
@@ -1,27 +1,71 @@
+            function(response) {
+                blade.currentEntity.errors = [];
+
+                if (response.passwordIsValid) {
```
This long switch/case can be simplified with a one-line loop where we can use JS `Object.keys()` to get properties names and use them to choose appropriate translation key from resources

avas commented:
Pushed some changes to ask users to change their password after sign in:
- `AccountEntity`: added persistable `PasswordExpired` property to indicate whether user should be prompted to change their password on next sign-in. Value of this property is passed to `ApplicationUserExtended.PasswordExpired`.
- `SecurityService`
    - `ResetPasswordAsync`: added `forcePasswordChange` parameter that gets assigned to `AccountEntity.PasswordExpired`.
    - `ChangePasswordAsync`, `ResetPasswordAsync`: added code to reset `PasswordExpired` flag for user (since the previous expired password *did* change, and the new password is not expired anymore).
- `SecurityController`: added `ResetCurrentUserPassword` method that resets password of current platform user. It works same as `ResetPassword` method, but does not require `PredefinedPermissions.SecurityUpdate` permission and can be used by any authorized user.
- Changed `changePasswordDialog`: now it changes password for current user only (previous implementation have been able to change password for any user). This was necessary because previous implementation required admin privileges and thus cannot be used by regular users to change their password. The change was suggested by @tatarincev.
- Extracted password validation code from `account-resetPassword` blade to the new `platformWebApp.passwordValidationService` to reduce code duplication.
- Added password strength validation to `changePasswordDialog`.
- `app.js`: added code that redirects user to `passwordChangeDialog` after sign in if they have to change their password.

tatarincev commented on VirtoCommerce.Platform.Web/Scripts/app/security/security.js:
```diff
@@ -115,18 +115,33 @@
+                            function (result) {
+                                $scope.passwordValidationResult = result;
+
+                                return result.passwordIsValid ? $q.resolve() : $q.reject();
```
Why do you use `$q.resolve() : $q.reject();` this? I think all can work without these calls. Just returns result.passwordIsValid  should be enought.

avas replied:
I did it because of `ui-validate-async` nature - it [expects](https://github.com/angular-ui/ui-validate/blob/master/src/validate.js#L108) to receive one of the following objects:
* Promise - this way control will be valid if the promise was resolved, and invalid if the promise was rejected;
* Anything else - this way `uiValidateAsync` directive will wrap it to promise and resolve/reject it if the value can or can't be implicitly converted to `true`.

If we'll replace `return result.passwordIsValid ? $q.resolve() : $q.reject()` with just `return result.passwordIsValid`, `validatePasswordAsync()` method will always return resolved promise with `passwordIsValid` as its result, because the HTTP call to password validation method always returns `Ok()`, even if password is invalid. So the password input will always stay valid, because the `ui-validate-async` directive looks only for resolution/rejection of the promise and ignores resolution result.

tatarincev replied:
If this  function returns `result.passwordIsvalid` boolean value which  will normally be handled by `ui-validate-async`  directive in else statement  https://github.com/angular-ui/ui-validate/blob/master/src/validate.js#L111

avas replied:
I tried that, and the password input always stays valid. Apparently, it happens because `validatePasswordAsync()` still returns promise, so it gets handled by `if` statement on [line 108](https://github.com/angular-ui/ui-validate/blob/master/src/validate.js#L108).

tatarincev commented on VirtoCommerce.Platform.Web/Scripts/app/security/blades/account-resetPassword.js:
```diff
@@ -1,27 +1,37 @@
+            function(result) {
+                angular.extend(blade.currentEntity, result);
+
+                return result.passwordIsValid ? $q.resolve() : $q.reject();
```
Why do you use $q.resolve() : $q.reject(); this? I think all can work without these calls. Just returns result.passwordIsValid should be enought.

avas commented:
Added password validation to account creation blade:
![image](https://user-images.githubusercontent.com/1835759/45490562-a6b7bb80-b791-11e8-8ba6-0d3d41d2bda2.png)

avas commented:
* `passwordValidationService`: added throttling for password validation requests. Default throttling interval is set to 100 ms, but this value can be changed if needed by changing `passwordValidationService.throttleTimeoutMilliseconds` value.
* Added `va-password-input` directive to make password validation easier. It can be used like this:

```html
<label class="form-label">{{ 'platform.blades.change-password.password' | translate }}</label>
<va-password-input password-placeholder="{{ 'platform.blades.account-resetPassword.placeholders.new-password' | translate }}"
                   password-too-weak-message="{{ 'platform.blades.account-resetPassword.validations.password-too-weak' | translate }}"
                   run-password-validation="validatePasswordAsync(value)"
                   new-password="password" />
```

It accepts following parameters:
* `password-placeholder`: placeholder text for password input;
* `password-too-weak-message`: message that will be displayed if password is considered too weak. It will be followed by error messages;
* `run-password-validation`: method binding that will be used for actual password validation. It must accept one parameter named `value` that will contain the entered password. It must return a promise that resolves to a result with following properties:
    * `passwordIsValid`: boolean value indicating if password is considered valid;
    * `errors`: array of resource keys that describe errors found when validating password;
    * If error resource strings contain any substitutable variables (e.g. `Passwords must be {{minPasswordLength}} or more characters long`), result must also contain those variables.
* `new-password`: binding to a property that will be used to store entered password value. Property value will be updated only if entered password is considered valid.
